### DISCUSSION TITLE
Panic when running as windows service: no console

### DISCUSSION THREAD

**Discussion by manoelmenezes** [seaweedfs/seaweedfs#2865] [Q&A]
When running as windows service it fails:

```
panic: no console

goroutine 1 [running]:
modernc.org/libc.newFile(0x1bdb900, 0x0)
	/go/pkg/mod/modernc.org/libc@v1.9.5/libc_windows.go:227 +0xb8
modernc.org/libc.init()
	/go/pkg/mod/modernc.org/libc@v1.9.5/libc.go:53 +0x2751
```

**Comment by chrislusf** [comment]
Seems a generic problem and not specific to SeaweedFS.

**Comment by woodenphone** [comment]
I've reproduced this type of error on a Windows 10 VM using the latest release [windows_amd64_large_disk.zip](https://github.com/chrislusf/seaweedfs/releases/download/2.96/windows_amd64_large_disk.zip)

My hunch is it's related to something relating to how windows services work with STDIN, STDOUT, STDERR, or whatever logging framework(s) windows likes to use for services.

The only places that gave me any kind of useful information about what my seaweedfs service was doing were the seaweedfs log files and the redirected STDERR/STDOUT (Note that directing STDIN/STDERR/STDOUT to files may be part of the cause in my case).

At a quick look the problem can be bypassed by explicitly telling seaweedfs to log to file with the  `<executable> <logging_options> <command> <arguments>` invocation args pattern.

e.g. `C:\weed\ > C:\weed\weed.exe -logdir=C:\weed server -master.port=9333 -volume.port=8080 -dir="./data"`

Here is a copy of the more pertinent notes I made while reproducing this:

# Running seaweedfs on windows as a service
"It works for me" (2022-04-01)
* https://stackoverflow.com/questions/3582108/create-windows-service-from-executable

* http://nssm.cc/download

Created "C:\weed"

Gave everyone full write/modify/read/etc. permissions on "C:\weed" via rightclick>properties>security

Weed binary copied to: "C:\weed\weed.exe"

NSSM copied to: "C:\weed\nssm.exe"

Admin PowerShell:
```
cd C:\weed\
C:\weed\nssm.exe install weed
```

In nssm GUI:

NSSM GUI > Application
path: `C:\weed\weed.exe`
startup directory: `C:\weed\`
args: `-logdir=C:\weed server -master.port=9333 -volume.port=8080 -dir="./data" -filer.port=8888`

Click "install service" button

Admin powershell:
```
PS C:\weed>

## Start service
sc.exe start weed

## Peek at service properties:
$service = Get-WmiObject -Class Win32_Service -Filter "Name='weed'"
$service
$service | Get-Member
```

Remove service before recreating:
```
sc.exe stop "weed"
$service = Get-WmiObject -Class Win32_Service -Filter "Name='weed'"
$service.delete()
```

NSSM GUI > I/O
input (stdin): `C:\weed\weed.stdin`
output (stdout): `C:\weed\weed.stdout`
error (srderr): `C:\weed\weed.stderr`
* If I set these it causes the error
```
panic: no console

goroutine 1 [running]:
modernc.org/libc.newFile(0x26c5480?, 0x0)
	/go/pkg/mod/modernc.org/libc@v1.14.12/libc_windows.go:293 +0xb8
modernc.org/libc.init()
	/go/pkg/mod/modernc.org/libc@v1.14.12/libc.go:56 +0x3831
```

**Reply by manoelmenezes** [reply]
Thanks for the deep dive.
I was already starting weed.exe passing the log dir as parameter, and that error happened. So I think the problem is still not solved:

```
weed.exe -logdir="C:\Users\Administrator\Documents\work\logs" server -resumeState -ip="127.0.0.1" -dataCenter='dc1' -master -master.volumeSizeLimitMB=1024 -master.port=9333 -master.dir="C:\Users\Administrator\Documents\work\object_store\master" -volume -volume.max=1024 -volume.port=8080 -dir="C:\Users\Administrator\Documents\work\object_store\volume" -volume.index='leveldb' -filer -filer.port=8888 -filer.dirListLimit=100000 -s3 -s3.port=4572
```

I found one alternative that worked which was creating a command file (.cmd) and using this file when creating the windows service

`weed.cmd`
```
@echo off
weed.exe -logdir="C:\Users\Administrator\Documents\work\logs" server -resumeState -ip="127.0.0.1" -dataCenter='dc1' -master -master.volumeSizeLimitMB=1024 -master.port=9333 -master.dir="C:\Users\Administrator\Documents\work\object_store\master" -volume -volume.max=1024 -volume.port=8080 -dir="C:\Users\Administrator\Documents\work\object_store\volume" -volume.index='leveldb' -filer -filer.port=8888 -filer.dirListLimit=100000 -s3 -s3.port=4572
```
